vkdevicechooser: init at 1.1

sigmike opened the pull request:
## Description of changes

Adds a new package `vkdevicechooser` that allows a user to select which vulkan device is used by setting environment variables. See https://github.com/aejsmith/vkdevicechooser.

The package requires the generated headers from the `vulkan-validation-layers` package, so I added them to the install phase.

See the discussion at https://discourse.nixos.org/t/building-vkdevicechooser-missing-vk-layer-dispatch-table-h/45196.

In the new package I mostly copied code from another package that adds a vulkan layer (I don't remember which one, but most likely `vulkan-extension-layer` or `vulkan-tools-lunarg`).

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sigmike committed:
vkdevicechooser: init at 1.1

Inspired by the Arch package[1], it uses a fork of vkdevicechooser that
uses the latest version of the Vulkan libraries and doesn't require the
generated headers from vulkan-validation-layers[2].

Related discussion:
https://discourse.nixos.org/t/building-vkdevicechooser-missing-vk-layer-dispatch-table-h/45196

[1] https://aur.archlinux.org/packages/vkdevicechooser
[2] https://github.com/jiriks74/vkdevicechooser/commit/2eecb26bcc1a61bed8592a324a71a93443d533df

sigmike committed:
maintainers: add sigmike

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/building-vkdevicechooser-missing-vk-layer-dispatch-table-h/45196/4

AndersonTorres commented on pkgs/development/tools/vulkan-validation-layers/default.nix:
```diff
@@ -60,6 +60,16 @@ stdenv.mkDerivation rec {
     "-Wno-dev"
   ];
 
+  installPhase = ''
```
If I understood correctly, you are rewriting the whole phase because it does not install certain files.

This being the case, you should put the extra installation under `postInstall` instead.

sigmike replied:
Indeed. I did that, thank you for the suggestion.

AndersonTorres commented on pkgs/by-name/vk/vkdevicechooser/package.nix:
```diff
@@ -0,0 +1,58 @@
+    done
+  '';
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

sigmike replied:
I applied these changes.

AndersonTorres commented on pkgs/by-name/vk/vkdevicechooser/package.nix:
```diff
@@ -0,0 +1,58 @@
+  meta = with lib; {
+    description = "Vulkan layer to force a specific device to be used";
+    homepage = "https://github.com/aejsmith/vkdevicechooser";
+    platforms = platforms.unix;
+    license = licenses.mit;
+    maintainers = [ maintainers.sigmike ];
```
```suggestion
    platforms = lib.platforms.unix;
    license = lib.licenses.mit;
    maintainers = [ lib.maintainers.sigmike ];
```

AndersonTorres commented on pkgs/by-name/vk/vkdevicechooser/package.nix:
```diff
@@ -0,0 +1,58 @@
+    addToSearchPath XDG_DATA_DIRS @out@/share
+  '';
+
+  mesonFlags = [ "-Dc_args=-I${vulkan-validation-layers}/include" ];
```
```suggestion
  mesonFlags = [ 
    (lib.mesonOption "c_args" "-I${lib.getInclude vulkan-validation-layers}/include")
  ];
```

P.S.: do not forget to run formatter.

Ralith commented:
> The package requires the generated headers from the vulkan-validation-layers package, so I added them to the install phase.

Have you investigated why upstream doesn't install these by default?

sigmike commented:
> Have you investigated why upstream doesn't install these by default?

I haven't. But I just looked at the Arch package and discovered they actually use a fork that doesn't require this generated header and uses the one provided by `vulkan-utility-libraries`. So this change to `vulkan-validation-layers` is no longer necessary. I made the change in 6756372f68c54c085b7f274ddbc106e89d874963.

sigmike commented:
I applied the change suggested by @FliegendeWurst, and squashed the commits of version 1.0.

I also rebased from master, but that made the "Compare" link on GitHub unreadable, so I rebased back from the original commit. You can see the final "Compare" here: [6756372..aeb0565](https://github.com/NixOS/nixpkgs/compare/6756372f68c54c085b7f274ddbc106e89d874963..aeb0565efec03837e0b7436ac2b56287e07f68cb).

Ralith commented on pkgs/by-name/vk/vkdevicechooser/package.nix:
```diff
@@ -0,0 +1,56 @@
+    vulkan-utility-libraries
+  ];
+
+  # Help vulkan-loader find the validation layers
```
"the validation layers" is incorrect here.

sigmike replied:
I changed that to "the layer".

FliegendeWurst reviewed (approved):
Builds on x64 Linux, did not test functionality

FliegendeWurst commented:
By the way, if you add your git email to Github, the commits show up as yours. It applies retroactively too.
